### ISSUE TITLE
iPad Pro Smart Keyboard folio does not work with new release

### ISSUE THREAD

**Issue by kaiserlxxii** [moonlight-stream/moonlight-ios#404]
Just to make sure it was not on my end, I restarted the iPad, killed and relaunched the app, restarted the remote computer, and logged into the VNC app to verify the keyboard works remotely with other apps.  

Since the new version was released, the keyboard folio does not work. This is not a Bluetooth keyboard but a smart connect keyboard.

**Comment by cgutman** [comment]
What version of iOS are you running?

Do you have a Bluetooth keyboard you can try?

If you 3 finger tap (which would normally display the software keyboard), does the folio keyboard start working then?

Does the software keyboard work if you 3 finger tap without the folio keyboard connected?

Do you have a Mac? Do you know how to build Moonlight from source (it's fine if not)?

**Comment by ascagnel** [comment]
The folio keyboard works if you three-finger tap; however, it’s not ideal.

- the command keys are not passed through to the host system (they’re usually sent to the host system as `VK_LWIN` and `VK_RWIN` for the left and right keys, respectively)
- similarly, the option/alt keys and control key are not passed into the host system

**Comment by cgutman** [comment]
@ascagnel I don't think any of that sounds like a regression. That's how keyboard support worked via the standard software IME and keyboards prior to iOS 13.4. It seems like Folio is not treated as a standard keyboard, but as an extension to the software IME. That's why you don't get the full support for key-up and key-down events on iOS 13.4.

**Comment by ascagnel** [comment]
Would this then be considered a new feature?  I’ve used a few other apps that have been updated for the iPadOS 13.4 keyboard & mouse support, and they’re able to do more with the existing folio and a bluetooth mouse is able to (including passing special keys and some degree of mouse control to the host PC).

**Comment by cgutman** [comment]
I'm not really sure. @kaiserlxxii seemed to imply this was a regression with the Folio on v4.0.0 which is why I'm trying to sort out what's going on.

Mouse support is tracked by a separate issue #402 

The Folio should support modifier keys, but only in combination with another non-modifier key. This is because the IME has very limited support for modifiers by creating `UIKeyCommands` for every possible key combination. See https://github.com/moonlight-stream/moonlight-ios/blob/master/Limelight/Input/TextFieldKeyboardDelegate.m#L93

But again, none of this is an issue with a real Bluetooth HID keyboard on iOS 13.4 which has full support for independent key-up and key-down events and real HID scancodes. See https://github.com/moonlight-stream/moonlight-ios/commit/75ab28a2c3634fe1975ab5f96559da6097f34c09

Does that behavior match what you see? Can you try a real Bluetooth keyboard to compare the functionality?

There's nothing we can do if the new keyboard support in iOS 13.4 doesn't support the Folio and it continues to use the legacy IME codepath. That's something Apple will have to fix.

**Comment by samwelnella** [comment]
I have a Smart Keyboard folio and it’s working fine. No 3 finger tap needed.

**Comment by kaiserlxxii** [comment]
Sorry, I’ve been pretty sick so I became unresponsive.  Whatever suggested the 3 finger tap fixed the issue, but this is an issue, and should not be required for any smart connect keyboard (in my opinion).

So everyone is aware, I use this app every day, and once it upgraded to the newest version, it stopped working (or in reality, caused me to have to 3 finger tap to get it to work).

So everyone is also aware (I apologize about being all over the place but I’m just scanning each comment and answering as I see them), this is an apple product and was released in conjunction with the 2018 iPad Pro so it is still supported by the current and all future (for the foreseeable future) iOS updates.

I don’t have a Bluetooth keyboard so I cant test it unfortunately.

Finally, no I don’t have a Mac, and no I don’t know how to build Moonlight from source.  I just saw a feature stop working once the new update went through, and so I felt I should let someone know.

**Comment by cgutman** [comment]
> Sorry, I’ve been pretty sick so I became unresponsive. Whatever suggested the 3 finger tap fixed the issue, but this is an issue, and should not be required for any smart connect keyboard (in my opinion).

No problem, hopefully you are feeling better.

The 3 finger tap is required to activate the text input field that triggers the software keyboard (otherwise it would be displayed all the time and obstruct the stream itself). In your case, the software keyboard doesn't actually display due to your attached physical keyboard, but there's no reliable way for Moonlight to know that you have a keyboard connected. As a result, we must not activate our text input control until a user gesture triggers it.

This limitation doesn't apply for HID keyboards on iOS 13.4 which can provide input at any time regardless of the state of the software keyboard (and also have other benefits, like being able to send modifiers keys alone and providing separate key-up and key-down events).

> So everyone is aware, I use this app every day, and once it upgraded to the newest version, it stopped working (or in reality, caused me to have to 3 finger tap to get it to work).

OK, so it sounds like it regressed for you in v4.0.0. @ascagnel is seeing behavior like yours, but it's unclear if it's new regression for him or just behaving like it did prior to the keyboard updates in v4.0.0. @samwelnella is seeing the expected behavior on iOS 13.4 and later where the keyboard always works and passes modifiers. Something is clearly very different about your setups.

> So everyone is also aware (I apologize about being all over the place but I’m just scanning each comment and answering as I see them), this is an apple product and was released in conjunction with the 2018 iPad Pro so it is still supported by the current and all future (for the foreseeable future) iOS updates.

What version of iOS is your iPad Pro running?

Just so it's clear what products we're all talking about, can each of you post a link from Apple or Amazon to the exact keyboard you have?

> I don’t have a Bluetooth keyboard so I cant test it unfortunately.
> 
> Finally, no I don’t have a Mac, and no I don’t know how to build Moonlight from source. I just saw a feature stop working once the new update went through, and so I felt I should let someone know.

OK, no problem. Thanks for the report.

**Comment by ascagnel** [comment]
Quick follow-up:

- I'm using an 11" iPad Pro with the [Apple keyboard folio](https://www.apple.com/shop/product/MXNK2LL/A/smart-keyboard-folio-for-ipad-pro-11%E2%80%91inch-2nd-generation-us-english?fnode=37).
- I observed this behavior on iPadOS 13.4.1.

I believe keyboard input worked some time ago, before the keyboard input update in 13.4 (when the OS sent the input character rather than the key up/key down codes).

**Comment by cgutman** [comment]
@ascagnel Thanks, can you try to checkout and build the code at `eceb23e1cd11134f4d91c0c250a438109d70c831`? If that works again, that will tell us it's related to the new iOS 13.4 keyboard support.

If that still doesn't work, try going all the way back to `47aaa02d08248cda8677760e64fad09d56095c29` which was the last public release before v4.0.0. If that still doesn't work, we could be dealing with a regression in iOS 13.4 itself.

**Comment by cgutman** [comment]
Never mind, I was able to reproduce it with my old iPhone that I hadn't updated to iOS 13.4 yet. It turns out that the commit that introduced the regression was https://github.com/moonlight-stream/moonlight-ios/commit/69d1ff15a1399d0afe43026dbca39cc14bff1022

I still have no idea why that had a behavior change. It was intended to only clean up the code a bit and not change the behavior. I've reverted it and submitted a new v5.0.1 build to Apple for review.

Fixed by https://github.com/moonlight-stream/moonlight-ios/commit/5abad38956b46e02a3023fbc08fa237e7c00bc0a
